src/ast/passes/codegen_llvm.cpp

```diff
@@ -151,16 +151,90 @@ void CodegenLLVM::kstack_ustack(const std::string &ident,
+      module_->getContext(), "stack_scratch_failure", parent);
+  BasicBlock *merge_block = BasicBlock::Create(module_->getContext(),
+                                               "merge_block",
+                                               parent);
+
+  Value *stack_trace = b_.CreateGetStackScratchMap(stack_type,
+                                                   stack_scratch_failure,
+                                                   loc);
+
+  BasicBlock *get_stack_success = BasicBlock::Create(module_->getContext(),
+                                                     "get_stack_success",
+                                                     parent);
+  BasicBlock *get_stack_fail = BasicBlock::Create(module_->getContext(),
+                                                  "get_stack_fail",
+                                                  parent);
+  auto uint64_size = sizeof(uint64_t);
+
+  Value *ptr_to_stack_trace = b_.CreatePointerCast(stack_trace,
+                                                   b_.GET_PTR_TY());
+  Value *stack_size = b_.CreateGetStack(
+      ctx_, is_ustack, ptr_to_stack_trace, stack_type, loc);
+  Value *condition = b_.CreateICmpSGE(stack_size, b_.getInt32(uint64_size));
+  b_.CreateCondBr(condition, get_stack_success, get_stack_fail);
+
+  b_.SetInsertPoint(get_stack_fail);
+  if (bpftrace_.debug_output_)
+    b_.CreateDebugOutput("Failed to get stack. Error: %d",
+                         std::vector<Value *>{ stack_size },
+                         loc);
+  b_.CreateStore(b_.getInt64(0), stackid);
+  b_.CreateBr(merge_block);
+  b_.SetInsertPoint(get_stack_success);
+
+  Value *stack_len = b_.CreateUDiv(stack_size, b_.getInt32(uint64_size));
+  // uint64 *stack, uint64 len, uint64 seed
+  AllocaInst *seed = b_.CreateAllocaBPF(CreateUInt64(), "seed");
+  // use the pid as the seed if we're getting user stacks
+  if (is_ustack) {
+    b_.CreateStore(b_.CreateGetPidTgid(loc), seed);
+  } else {
+    b_.CreateStore(b_.getInt64(1), seed);
+  }
+
+  // LLVM-12 produces code that fails the BPF verifier because it
+  // can't determine the bounds of stack_len. The only thing that seems
+  // to work is truncating the type, which is fine because 255 is long enough.
+  Value *trunc_stack_len = b_.CreateTrunc(stack_len, b_.getInt8Ty());
+  Value *get_stack_id = b_.CreateCall(get_stack_id_func_,
+                                      { ptr_to_stack_trace,
+                                        trunc_stack_len,
+                                        b_.CreateLoad(b_.getInt64Ty(), seed) },
+                                      "get_stack_id");
+  b_.CreateDebugOutput("Stack id %llu. Len: %llu",
+                       std::vector<Value *>{ get_stack_id, stack_len },
+                       loc);
+
+  // Add the stack and id to the stack map
+  b_.CreateStore(get_stack_id, stackid);
+  b_.CreateMapUpdateElem(ctx_, stack_type.name(), stackid, stack_trace, loc);
```

Comment:
This is a good question. I think it would involve storing the stack length along with the stack, checking if the id exists, and seeing if the lengths don't match - or the more expensive thing where we loop over each address in both stacks.

Maybe we can check if debug mode is enabled.